presto-main/src/main/java/com/facebook/presto/catalogserver/CatalogServer.java

```diff
@@ -163,4 +256,45 @@ private static String writeValueAsString(Object value, ObjectMapper objectMapper
             throw new RuntimeException(e);
         }
     }
+
+    private static class CacheKey<T>
+    {
+        private final SessionRepresentation session;
+        private final T key;
+
+        private CacheKey(SessionRepresentation session, T key)
+        {
+            this.session = session;
+            this.key = key;
+        }
+
+        public SessionRepresentation getSession()
+        {
+            return session;
+        }
+
+        public T getKey()
+        {
+            return key;
+        }
+
+        @Override
```

Comment:
Can we also provide comment here to explain why we only compare the `key`, similar to https://github.com/prestodb/presto/blob/master/presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java#L1032-L1034